lib/model/autocomplete.dart

```diff
@@ -317,110 +413,61 @@ class MentionAutocompleteView extends ChangeNotifier {
+  final String raw;
+  final List<String> _lowercaseWords;
 
-  MentionAutocompleteQuery? get query => _query;
-  MentionAutocompleteQuery? _query;
-  set query(MentionAutocompleteQuery? query) {
-    _query = query;
-    if (query != null) {
-      _startSearch(query);
-    }
-  }
+  AutocompleteQuery(this.raw) : _lowercaseWords = raw.toLowerCase().split(' ');
 
-  /// Called when the app is reassembled during debugging, e.g. for hot reload.
+  /// Whether all of this query's words have matches in [words] that appear in order.
   ///
-  /// This will redo the search from scratch for the current query, if any.
-  void reassemble() {
-    if (_query != null) {
-      _startSearch(_query!);
-    }
-  }
-
-  Iterable<MentionAutocompleteResult> get results => _results;
-  List<MentionAutocompleteResult> _results = [];
-
-  Future<void> _startSearch(MentionAutocompleteQuery query) async {
-    final newResults = await _computeResults(query);
-    if (newResults == null) {
-      // Query was old; new search is in progress. Or, no listeners to notify.
-      return;
-    }
-
-    _results = newResults;
-    notifyListeners();
-  }
-
-  Future<List<MentionAutocompleteResult>?> _computeResults(MentionAutocompleteQuery query) async {
-    final List<MentionAutocompleteResult> results = [];
-    final iterator = sortedUsers.iterator;
-    bool isDone = false;
-    while (!isDone) {
-      // CPU perf: End this task; enqueue a new one for resuming this work
-      await Future(() {});
-
-      if (query != _query || !hasListeners) { // false if [dispose] has been called.
-        return null;
+  /// A "match" means the word in [words] starts with the query word.
+  bool _testContainsQueryWords(List<String> words) {
+    // TODO(#237) test with diacritics stripped, where appropriate
+    int wordsIndex = 0;
+    int queryWordsIndex = 0;
+    while (true) {
+      if (queryWordsIndex == _lowercaseWords.length) {
+        return true;
+      }
+      if (wordsIndex == words.length) {
+        return false;
       }
 
-      for (int i = 0; i < 1000; i++) {
-        if (!iterator.moveNext()) {
-          isDone = true;
-          break;
-        }
-
-        final User user = iterator.current;
-        if (query.testUser(user, store.autocompleteViewManager.autocompleteDataCache)) {
-          results.add(UserMentionAutocompleteResult(userId: user.userId));
-        }
+      if (words[wordsIndex].startsWith(_lowercaseWords[queryWordsIndex])) {
+        queryWordsIndex++;
       }
+      wordsIndex++;
     }
-    return results;
   }
-}
 
-class MentionAutocompleteQuery {
-  MentionAutocompleteQuery(this.raw, {this.silent = false})
-    : _lowercaseWords = raw.toLowerCase().split(' ');
+  @override
+  String toString() {
+    return '${objectRuntimeType(this, 'AutocompleteQuery')}(raw: $raw})';
+  }
 
-  final String raw;
+  @override
+  bool operator ==(Object other) {
+    return other is AutocompleteQuery && other.raw == raw;
```

Comment:
This is wrong — for example it would let queries of two different concrete subclasses be equal just because `raw` was equal.

Probably this override just shouldn't be here. It can't ever get invoked, right? Both concrete subclasses override it.

Similarly the toString and hashCode seem like they can't get invoked and so should just be dropped.